independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java

```diff
@@ -75,105 +78,128 @@ Collection<Resource> generate(DotName scope) {
+            ResultHandle updater = staticConstructor.invokeStaticMethod(newLockUpdater,
+                    staticConstructor.loadClass(contextInstances.getClassName()),
+                    staticConstructor.loadClass(Lock.class),
+                    staticConstructor.load(fields.lock.getName()));
+            staticConstructor.writeStaticField(fields.lockUpdater, updater);
+        }
+        staticConstructor.returnVoid();
+    }
+
+    private static Map<String, MethodDescriptor> implementLazyLocks(ClassCreator contextInstances,
+            Map<String, InstanceAndLock> idToFields) {
+        MethodDescriptor atomicReferenceFieldUpdaterCompareAndSet = MethodDescriptor.ofMethod(
+                AtomicReferenceFieldUpdater.class, "compareAndSet",
+                boolean.class, Object.class, Object.class, Object.class);
+        // generated lazy lock initializers methods
+        // private Lock lazy1l() {
+        //   Lock lock = this.1l;
+        //   if (lock != null) {
+        //     return lock;
+        //   }
+        //   lock = new ReentrantLock();
+        //   if (LAZY_1L_UPDATER.compareAndSet(this, null, lock)) {
+        //       return lock;
+        //   }
+        //   return this.1l;
+        // }
+        Map<String, MethodDescriptor> lazyLockMethods = new HashMap<>(idToFields.size());
+        for (var namedFields : idToFields.entrySet()) {
+            var fields = namedFields.getValue();
+            MethodCreator lazyLockMethod = contextInstances.getMethodCreator("lazy" + fields.lock.getName(), Lock.class)
+                    .setModifiers(ACC_PRIVATE);
+            ResultHandle lock = lazyLockMethod.readInstanceField(fields.lock, lazyLockMethod.getThis());
+            lazyLockMethod
+                    .ifNotNull(lock).trueBranch()
+                    .returnValue(lock);
+            ResultHandle newLock = lazyLockMethod.newInstance(MethodDescriptor.ofConstructor(ReentrantLock.class));
+            ResultHandle updated = lazyLockMethod.invokeVirtualMethod(atomicReferenceFieldUpdaterCompareAndSet,
+                    lazyLockMethod.readStaticField(fields.lockUpdater), lazyLockMethod.getThis(),
+                    lazyLockMethod.loadNull(), newLock);
+            lazyLockMethod
+                    .ifTrue(updated).trueBranch()
+                    .returnValue(newLock);
+            lazyLockMethod.returnValue(lazyLockMethod.readInstanceField(fields.lock, lazyLockMethod.getThis()));
+            lazyLockMethods.put(namedFields.getKey(), lazyLockMethod.getMethodDescriptor());
+        }
+        return lazyLockMethods;
+    }
+
     private void implementGetAllPresent(ClassCreator contextInstances, Map<String, InstanceAndLock> idToFields) {
         MethodCreator getAllPresent = contextInstances.getMethodCreator("getAllPresent", Set.class)
                 .setModifiers(ACC_PUBLIC);
-        // this.lockAll();
         // ContextInstanceHandle<?> copy1 = this.1;
```

Comment:
I've removed the lockAll/unlockAll here, since it doesn't appear for them to be useful @Ladicek @mkouba

Reply:
So the original idea was to get the snapshot at the given time, i.e. disallow writing during the invocation of this method but you're right that it's not very useful. ​The truth is also that this method is almost never used. And so +1 ;-).